### DISCUSSION TITLE
How to change some Jar version in Keycloak Quarkus

### DISCUSSION THREAD

**Discussion by ajaiswal595** [keycloak/keycloak#16409] [Q&A]
I'm trying to change mysql.mysql-connector-java-8.0.30.jar to mysql-connector-java-5.1.47.jar

So I deleted mysql.mysql-connector-java-8.0.30.jar from keycloak-20.0.1\lib\lib\main and placed mysql-connector-java-5.1.47.jar 

When I started Keycloak I got the error

`ERROR: Failed to run 'build' command.
ERROR: java.lang.IllegalArgumentException: E:\keycloak-20.0.1\lib\lib\main\mysql.mysql-connector-java-8.0.30.jar does not exist
ERROR: E:\keycloak-20.0.1\lib\lib\main\main\mysql.mysql-connector-java-8.0.30.jar does not exist
ERROR: E:\keycloak-20.0.1\lib\lib\main\mysql.mysql-connector-java-8.0.30.jar
For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.`

**Also, in the future, if we get a vulnerability in some jar and the solution is to upgrade that jar, how do we achieve that?**

As It is looking for a specific jar to be available in the lib folder it will not let us change any jar.

**Comment by ajaiswal595** [comment]
@sschu @pedroigor  any solution for this?

**Reply by sschu** [reply]
I think it is very unlikely that a different major version is actually compatible, at least if semantic versioning is used. So I don't think what you are trying to do works. For minor or patch version changes, it probably does work. However, I don't know if it is possible to just do it in the provided distribution. You can definitely do it by building your own distribution with fixed versions of dependencies (and ideally provide a PR to also update upstream).

**Comment by pedroigor** [comment]
We are working on an installer tool to allow you to create your own distribution by removing and adding Quarkus dependencies and replacing dependencies.

We are not 100% sure yet how replacing dependencies will work without impacting supportability.

In your case, it works if you keep the same file name when replacing but the results are completely unexpected as those are the versions we support.

**Comment by pedroigor** [comment]
@nehachopra27 @alexeykazakov ^

**Comment by ajaiswal595** [comment]
If I change mysql-jdbc.version to 8.0.31 then my builds fails

I see lib/lib/main/

<img width="673" alt="image" src="https://user-images.githubusercontent.com/8745632/212234140-52766b01-3051-4779-a2f1-b8125d4ffae8.png">



So I'm trying to change the version of MySQL-connector 

<img width="1098" alt="image" src="https://user-images.githubusercontent.com/8745632/212233470-245a40c5-d888-4899-be0a-7896b8d28c0c.png">

Build fails, 

**I just tried to change the latest version of mysql connector from 8.0.30 to 8.0.31**

<img width="1666" alt="image" src="https://user-images.githubusercontent.com/8745632/212235030-c9ae9daf-8cd2-4344-80ec-d0529cd23018.png">


Where do we need to change the version of Jars?

**Comment by pedroigor** [comment]
Can we make a step back and understand first why you need to change the driver version?

**Reply by ajaiswal595** [reply]
Consider it just for example

![image](https://user-images.githubusercontent.com/8745632/212333302-5c9d996e-d884-4a00-b341-c36288215528.png)

If you see till 8.0.27 mysql-connector-jar is vulnerable and we never know when 8.0.30's vulnerability get detected and reported as vulnerable jar.

It's not just mysql-connector-jar but it can be any jar like netty, elytron, jboss-logmanager etc.

So to fix that vulnerability the community release with minor fix and we can use the latest jar where the vulnerability will be fixed.

In wildfly, we could do it by adding it to the modules folder directly.

How can I achieve this with the current distribution of quakus?

**Reply by pedroigor** [reply]
Try replacing the JAR by keeping the same name. But if the version is not compatible with the `quarkus-jdbc-mysql` you might have issues.

As I mentioned before, we are working to improve this kind of patching and how to properly handle CVEs in dependencies. Right now, updates to dependencies only in new releases.

**Reply by ajaiswal595** [reply]
On the same line here I'm not talking about the provider
Is there any option that we can load our custom jar (Like some utility) by setting the classpath in Quarkus/Keycloak?

**Reply by pedroigor** [reply]
There isn't. The only way to include JARs in the classpath (a flatclasspath btw) is droping JARs to `providers` dir.

**Reply by ajaiswal595** [reply]
Actually these days we identified vulnerabilities in sankeyml, undertow, logmanager  as it is very tight coupled with quarkus and  *.dat file we can not delete those jars and put them into the provider.

I'm hoping for some kind of tool that can modify the .dat file based on the provided jar version.

**Reply by ajaiswal595** [reply]
[https://github.com/keycloak/keycloak/discussions/16517](https://github.com/keycloak/keycloak/discussions/16517
)
